### PR TITLE
[msan] Add 32-bit platforms support

### PR DESCRIPTION
**Pull request by k-kashapov**
References https://github.com/llvm/llvm-project/issues/103057

Added `VAArgHelper` functions for platforms: ARM32, i386, RISC-V, PowerPC32, MIPS32.

ARM, RISCV and MIPS share similar conventions regarding va args. Therefore `VAArgGenericHelper` was introduced to avoid code duplication.

### CODE REVIEW

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 1]
```diff
@@ -4487,6 +4488,8 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
       if (EagerCheck) {
         insertShadowCheck(A, &CB);
         Size = DL.getTypeAllocSize(A->getType());
+        if (ArgOffset + Size > kParamTLSSize)
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:4491]: This requires a separate patch with a test

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 2]
```diff
@@ -1007,12 +1008,12 @@ void MemorySanitizer::initializeModule(Module &M) {
...
     case Triple::Linux:
       switch (TargetTriple.getArch()) {
-      case Triple::x86_64:
-        MapParams = Linux_X86_MemoryMapParams.bits64;
-        break;
       case Triple::x86:
         MapParams = Linux_X86_MemoryMapParams.bits32;
         break;
+      case Triple::x86_64:
+        MapParams = Linux_X86_MemoryMapParams.bits64;
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:1015]: All such optional cleanups should be in separate PRs.

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 3]
```diff
@@ -6057,21 +6201,57 @@ static VarArgHelper *CreateVarArgHelper(Function &Func, MemorySanitizer &Msan,
   // VarArg handling is only implemented on AMD64. False positives are possible
   // on other platforms.
   Triple TargetTriple(Func.getParent()->getTargetTriple());
-  if (TargetTriple.getArch() == Triple::x86_64)
+
+  if (TargetTriple.getArch() == Triple::x86) {
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:6205]: for each new case we need a test, please clone corresponding existing tests

**Comment by vitalybuka** [comment]
> Check code formatting / code_formatter (pull_request) 

Please fix^

**Comment by vitalybuka** [comment]
Please don't forget to click "re-request review" when ready

File: llvm/test/Instrumentation/MemorySanitizer/ARM32/vararg-arm32.ll [hunk 4]
```diff
@@ -0,0 +1,91 @@
+; RUN: opt < %s -S -passes=msan 2>&1 -msan-origin-base=0x40000000 -msan-and-mask=0x80000000 | FileCheck %s
+
+target datalayout = "E-m:m-i8:8:32-i16:16:32-i32:32-n32-S64"
+target triple = "arm--linux"
+
+define i32 @foo(i32 %guard, ...) {
```
Comment (vitalybuka) [inline llvm/test/Instrumentation/MemorySanitizer/ARM32/vararg-arm32.ll:6]: Assuming this one was copied from something else, 
I suggest this sequence of pull requests:
1. Move/copy files unchanged
2. switch "target tripple" to 32bit and update tests with invalid instrumentation (this step only if compiled does not crash or fail on assert)
3. This PR #109284
 a PR with moving copying files unchanged, and in the current PR "target triple"

Comment (k-kashapov) [reply llvm/test/Instrumentation/MemorySanitizer/ARM32/vararg-arm32.ll:6]: Thank you.
Yes, this was copied from mips64 and modified to fit the tests.
So, the first step is we open a PR with tests for all the 32bit platforms, but their "target triple" is their respective 64bit counterpart? Did I understand it right?

Comment (k-kashapov) [reply llvm/test/Instrumentation/MemorySanitizer/ARM32/vararg-arm32.ll:6]: Second step is also unclear:
If we just switch the target triple to 32bit, most of the checks will be invalid. Thus, "update tests with invalid instrumentation" would mean update all tests.
Also, since this PR is the last, MemorySanitizer will not be updated, so when running tests between 2nd and 3rd PR's the tests will fail with "unsupported architecture" assert.
I believe, this is not the intended course of action. Could you please, elaborate on the steps?

**Comment by k-kashapov** [comment]
Rebased to main to get all llvm/tests changes from other PR's

File: llvm/test/Instrumentation/MemorySanitizer/ARM32/vararg-arm32.ll [hunk 5]
```diff
@@ -1,8 +1,8 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
-; RUN: opt < %s -S -passes=msan 2>&1 | FileCheck %s
+; RUN: opt < %s -S -passes=msan -msan-origin-base=0x40000000 -msan-and-mask=0x80000000 2>&1 | FileCheck %s
```
Comment (vitalybuka) [inline llvm/test/Instrumentation/MemorySanitizer/ARM32/vararg-arm32.ll:2]: Can we remove those and add proper consts into msan pass?

File: llvm/test/Instrumentation/MemorySanitizer/i386/sse-intrinsics-i386.ll [hunk 6]
```diff
@@ -371,17 +411,19 @@ declare <4 x float> @llvm.x86.sse.rsqrt.ss(<4 x float>) nounwind readnone
 define void @test_x86_sse_stmxcsr(ptr %a0) #0 {
 ; CHECK-LABEL: @test_x86_sse_stmxcsr(
 ; CHECK-NEXT:    [[TMP1:%.*]] = load i64, ptr @__msan_param_tls, align 8
+; CHECK-NEXT:    [[TMP5:%.*]] = load i64, ptr @__msan_va_arg_overflow_size_tls, align 8
+; CHECK-NEXT:    [[TMP6:%.*]] = add i64 0, [[TMP5]]
 ; CHECK-NEXT:    call void @llvm.donothing()
 ; CHECK-NEXT:    [[TMP2:%.*]] = ptrtoint ptr [[A0:%.*]] to i64
```
Comment (vitalybuka) [inline llvm/test/Instrumentation/MemorySanitizer/i386/sse-intrinsics-i386.ll:417]: there are a lot of conversions from ptr to i64, which is too much
but if msan works as is, let's keep this for follow up patches.

File: llvm/test/Instrumentation/MemorySanitizer/i386/vararg-too-large.ll [hunk 7]
```diff
@@ -111,11 +113,28 @@ define dso_local i64 @many_args() {
 ; CHECK-NEXT:    store i64 0, ptr inttoptr (i64 add (i64 ptrtoint (ptr @__msan_param_tls to i64), i64 776) to ptr), align 8
 ; CHECK-NEXT:    store i64 0, ptr inttoptr (i64 add (i64 ptrtoint (ptr @__msan_param_tls to i64), i64 784) to ptr), align 8
 ; CHECK-NEXT:    store i64 0, ptr inttoptr (i64 add (i64 ptrtoint (ptr @__msan_param_tls to i64), i64 792) to ptr), align 8
+; CHECK-NEXT:    store i64 0, ptr @__msan_va_arg_tls, align 8
 ; CHECK-NEXT:    store i64 0, ptr inttoptr (i64 add (i64 ptrtoint (ptr @__msan_va_arg_tls to i64), i64 8) to ptr), align 8
 ; CHECK-NEXT:    store i64 0, ptr inttoptr (i64 add (i64 ptrtoint (ptr @__msan_va_arg_tls to i64), i64 16) to ptr), align 8
 ; CHECK-NEXT:    store i64 0, ptr inttoptr (i64 add (i64 ptrtoint (ptr @__msan_va_arg_tls to i64), i64 24) to ptr), align 8
 ; CHECK-NEXT:    store i64 0, ptr inttoptr (i64 add (i64 ptrtoint (ptr @__msan_va_arg_tls to i64), i64 32) to ptr), align 8
 ; CHECK-NEXT:    store i64 0, ptr inttoptr (i64 add (i64 ptrtoint (ptr @__msan_va_arg_tls to i64), i64 40) to ptr), align 8
+; CHECK-NEXT:    store i64 0, ptr inttoptr (i64 add (i64 ptrtoint (ptr @__msan_va_arg_tls to i64), i64 48) to ptr), align 8
```
Comment (vitalybuka) [inline llvm/test/Instrumentation/MemorySanitizer/i386/vararg-too-large.ll:122]: do you know why we get these new stores?

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 8]
```diff
@@ -5589,6 +5590,8 @@ struct VarArgAArch64Helper : public VarArgHelperBase {
 struct VarArgPowerPCHelper : public VarArgHelperBase {
   AllocaInst *VAArgTLSCopy = nullptr;
   Value *VAArgSize = nullptr;
+  const DataLayout &DL = F.getDataLayout();
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:5593]: please don't mixin unnecessary changes

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 9]
```diff
@@ -6004,37 +6024,152 @@ struct VarArgSystemZHelper : public VarArgHelperBase {
   }
 };
 
-/// MIPS-specific implementation of VarArgHelper.
-/// NOTE: This is also used for LoongArch64.
+/// i386-specific implementation of VarArgHelper.
+struct VarArgI386Helper : public VarArgHelperBase {
+  AllocaInst *VAArgTLSCopy = nullptr;
+  Value *VAArgSize = nullptr;
+  const DataLayout &DL = F.getDataLayout();
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:6031]: Please keep same approach 
to have those local

```
  const DataLayout &DL = F.getDataLayout();
  unsigned IntptrSize = DL.getTypeStoreSize(MS.IntptrTy);
```

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 10]
```diff
@@ -6004,37 +6024,152 @@ struct VarArgSystemZHelper : public VarArgHelperBase {
...
   VarArgGenericHelper(Function &F, MemorySanitizer &MS,
                       MemorySanitizerVisitor &MSV, const unsigned VAListTagSize)
       : VarArgHelperBase(F, MS, MSV, VAListTagSize) {}
 
   void visitCallBase(CallBase &CB, IRBuilder<> &IRB) override {
     unsigned VAArgOffset = 0;
-    const DataLayout &DL = F.getDataLayout();
-    for (Value *A :
-         llvm::drop_begin(CB.args(), CB.getFunctionType()->getNumParams())) {
+    for (const auto &[ArgNo, A] : llvm::enumerate(CB.args())) {
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:6153]: do we need to change the loop? drop_begin looks nicer

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 11]
```diff
@@ -6004,37 +6024,152 @@ struct VarArgSystemZHelper : public VarArgHelperBase {
...
       Value *Base = getShadowPtrForVAArgument(IRB, VAArgOffset, ArgSize);
       VAArgOffset += ArgSize;
-      VAArgOffset = alignTo(VAArgOffset, 8);
+      VAArgOffset = alignTo(VAArgOffset, IntptrSize);
       if (!Base)
         continue;
       IRB.CreateAlignedStore(MSV.getShadow(A), Base, kShadowTLSAlignment);
     }
 
-    Constant *TotalVAArgSize = ConstantInt::get(IRB.getInt64Ty(), VAArgOffset);
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:6037]: why do we need int64 here?

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 12]
```diff
@@ -6043,8 +6178,9 @@ struct VarArgGenericHelper : public VarArgHelperBase {
   void finalizeInstrumentation() override {
     assert(!VAArgSize && !VAArgTLSCopy &&
            "finalizeInstrumentation called twice");
+
     IRBuilder<> IRB(MSV.FnPrologueEnd);
-    VAArgSize = IRB.CreateLoad(IRB.getInt64Ty(), MS.VAArgOverflowSizeTLS);
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:6047]: can we change runtime to uptr?

File: llvm/test/Instrumentation/MemorySanitizer/i386/vararg-too-large.ll (hunk 7 above)
Comment (k-kashapov) [reply llvm/test/Instrumentation/MemorySanitizer/i386/vararg-too-large.ll:122]: I guess, this comes from [System V ABI 64-bit](https://cs61.seas.harvard.edu/site/pdf/x86-64-abi-20210928.pdf) part 3.5.7 and [System V ABI 32-bit](https://refspecs.linuxbase.org/elf/abi386-4.pdf) page 167:
* On x86_64 `va_list` is a struct and there is  dedicated `register_overflow_area`. This area is used to pass shadow for arguments that don't fit into registers. In IR we observe it as increase in offset against `__msan_va_arg_tls` for 6th argument.
* On i386 `va_list` is a `void *` and it does not imply any space separation for shadow of arguments that are passed in registers and in memory. This behavior is similar to `GenericHelper`.

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp (hunk 8 above)
Comment (k-kashapov) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:5593]: Sorry, will fix this now

File: llvm/test/Instrumentation/MemorySanitizer/ARM32/vararg-arm32.ll (hunk 5 above)
Comment (k-kashapov) [reply llvm/test/Instrumentation/MemorySanitizer/ARM32/vararg-arm32.ll:2]: We can, but this will require some research on where to put shadow and origin. Currently, I cannot say what their proper addresses will be.

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp (hunk 12 above)
Comment (k-kashapov) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:6047]: This comment is also unclear, could you specify your question please?

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp (hunk 1 above)
Comment (vitalybuka) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp]: I guess this piece is gone.

File: llvm/test/Instrumentation/MemorySanitizer/ARM32/vararg-arm32.ll (hunk 5 above)
Comment (vitalybuka) [reply llvm/test/Instrumentation/MemorySanitizer/ARM32/vararg-arm32.ll:2]: Then I afraid if someone decide to fix them will forget to update tests.
Can you put a comment in MemoryMapParams for corresponding platforms // FIXME: Remove -msan-origin-base -msan-and-mask added by #109284 after picking good constants.

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 13]
```diff
@@ -6004,8 +6026,122 @@ struct VarArgSystemZHelper : public VarArgHelperBase {
   }
 };
 
-/// MIPS-specific implementation of VarArgHelper.
-/// NOTE: This is also used for LoongArch64.
+/// i386-specific implementation of VarArgHelper.
+struct VarArgI386Helper : public VarArgHelperBase {
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:6030]: which one did you use as a source?

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 14]
```diff
@@ -5709,11 +5716,26 @@ struct VarArgPowerPCHelper : public VarArgHelperBase {
...
+        RegSaveAreaPtrPtr = IRB.CreateIntToPtr(
+            IRB.CreateAdd(IRB.CreatePtrToInt(VAListTag, MS.IntptrTy),
+                          ConstantInt::get(MS.IntptrTy, 8)),
+            MS.PtrTy);
+      }
+
       Value *RegSaveAreaPtr = IRB.CreateLoad(MS.PtrTy, RegSaveAreaPtrPtr);
       Value *RegSaveAreaShadowPtr, *RegSaveAreaOriginPtr;
-      const Align Alignment = Align(8);
+      const DataLayout &DL = F.getDataLayout();
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:5736]: We have DL declared above

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp (hunk 10 above)
Comment (vitalybuka) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp]: Actually this one looks like outlier ,others already use ArgNo style : )
Still better not to mix NFC and real changes.

Feel free to create a new PR to fix this inconsistency.

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp (hunk 8 above)
Comment (vitalybuka) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp]: Unresolved for follow up PRs.

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp (hunk 10 above)
Comment (vitalybuka) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp]: Unresolved for follow up PRs.

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp (hunk 14 above)
Comment (k-kashapov) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:5736]: DL is declared in `visitCallBase`, this funciton is `finalizeInstrumentation`.

**Review by vitalybuka — APPROVED** [review]
LGTM
Can merge after the rest of comments is resolved.

Though, I didn't rely validate vararg  ABI for new platforms.

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp (hunk 13 above)
Comment (k-kashapov) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:6030]: If I remember correctly, this has been copied from GenericHelper. Then `visitCallBase` has been modified to differentiate "by val" parameter passing.

File: llvm/test/Instrumentation/MemorySanitizer/i386/msan_i386_bts_asm.ll [hunk 15]
```diff
@@ -4,7 +4,7 @@
 ; RUN: opt < %s -msan-kernel=1 -msan-check-access-address=0 -msan-handle-asm-conservative=1 -S -passes=msan 2>&1 | FileCheck %s -check-prefixes=CONS
 
 target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
```
Comment (vitalybuka) [inline llvm/test/Instrumentation/MemorySanitizer/i386/msan_i386_bts_asm.ll:6]: I was fixing up few minor issue, and noticed that 
`target datalayout =` is still 64 bit, generating wrong IR

Could you please fix, and I'll merge tomorrow?

Comment (vitalybuka) [reply llvm/test/Instrumentation/MemorySanitizer/i386/msan_i386_bts_asm.ll:6]: I guess we can merge right now. Please create a new PR with updated datalayout and tests

Comment (k-kashapov) [reply llvm/test/Instrumentation/MemorySanitizer/i386/msan_i386_bts_asm.ll:6]: Thank you! I will open follow up PR's later today.

Comment (k-kashapov) [reply llvm/test/Instrumentation/MemorySanitizer/i386/msan_i386_bts_asm.ll:6]: Sorry for the delay. Didn't have time to open PR's yet. Will do this week.
